File: BEmuJava/bemu/src/main/java/com/bloomberglp/blpapi/ReferenceMessage.java
//------------------------------------------------------------------------------
// <copyright project="BEmu_maven" file="/BEmu_maven/bemu/src/main/java/com/bloomberglp/blpapi/ReferenceMessage.java" company="Jordan Robinson">
//     Copyright (c) 2013 Jordan Robinson. All rights reserved.
//
//     The use of this software is governed by the Microsoft Public License
//     which is included with this distribution.
// </copyright>
//------------------------------------------------------------------------------

package com.bloomberglp.blpapi;

import com.bloomberglp.blpapi.Name;
import com.bloomberglp.blpapi.Element;
import com.bloomberglp.blpapi.Message;
import com.bloomberglp.blpapi.CorrelationID;

import java.util.Map;
import java.util.List;
import java.util.ArrayList;

public class ReferenceMessage extends Message
{
    private final ReferenceElementSecurityDataArray _securities;

    ReferenceMessage(CorrelationID corr, Map<String, Map<String, Object>> fieldData) throws Exception
    {
    	super(new Name("ReferenceDataResponse"), corr, null);
        this._securities = new ReferenceElementSecurityDataArray(fieldData);
    }
    
    public int numElements()
    {
    	return 1;
    }
    
    List<Element> elements()
    {
    	List<Element> result = new ArrayList<Element>();
    	result.add(this._securities);
    	return result;
    }
    
    public Element getElement(String name) throws Exception
    {
    	if(name.equals(this._securities.name().toString()))
    		return this._securities;
    	
    	else
    		return super.getElement(name);
    }
    
    public String topicName()
    {
    	return "";
    }
    
    public Element asElement()
    {
    	return new ReferenceElementData(this);
    }
    
    protected StringBuilder prettyPrint(int tabIndent) throws Exception
    {
        StringBuilder result = new StringBuilder();

        result.append(String.format("fieldData[] = {%s", System.getProperty("line.separator")));
        result.append(this._securities.prettyPrint(1));
        result.append(String.format("}%s", System.getProperty("line.separator")));
        
        return result;
    }
    
}

File: BEmu/ReferenceDataRequest/ReferenceMessage.cs
//------------------------------------------------------------------------------
// <copyright project="BEmu_csh" file="ReferenceDataRequest/ReferenceMessage.cs" company="Jordan Robinson">
//     Copyright (c) 2013 Jordan Robinson. All rights reserved.
//
//     The use of this software is governed by the Microsoft Public License
//     which is included with this distribution.
// </copyright>
//------------------------------------------------------------------------------

namespace Bloomberglp.Blpapi.ReferenceDataRequest
{
    using System;
    using System.Collections.Generic;
    using System.Linq;
    using System.Text;

    internal class ReferenceMessage : Message
    {
        private readonly ReferenceElementSecurityDataArray _securities;

        internal ReferenceMessage(CorrelationID corr, Dictionary<string, Dictionary<string, object>> fieldData) : base(new Name("ReferenceDataResponse"), corr, null)
        {
            this._securities = new ReferenceElementSecurityDataArray(fieldData);
        }

        public override int NumElements { get { return 1; } }
        public override IEnumerable<Element> Elements { get { yield return this._securities; } }
        public override string TopicName { get { return ""; } }
        public override Element AsElement { get { return new ReferenceElement(this); } }

        public override object this[string name, int index]
        {
            get
            {
                ReferenceElementSecurityDataArray secArray = (ReferenceElementSecurityDataArray)this[name]; //parent class checks the name

                int itemCount = 0;
                foreach (var item in secArray.Elements)
                {
                    if (itemCount == index)
                        return item;

                    itemCount++;
                }

                return base[name, index];
            }
        }

        public override string ToString()
        {
            StringBuilder result = new StringBuilder();
            result.Append("ReferenceDataResponse (choice) = {" + Environment.NewLine);
            result.Append(this._securities.PrettyPrint(1));
            result.Append("}");

            return result.ToString();
        }

    }
}
